File: workflows/evaluator-optimizer/code/python/evaluator_optimizer.py
"""
Evaluator-Optimizer — Generate-evaluate feedback loop.

A generator produces output; an evaluator scores it and provides feedback;
the optimizer converts that feedback into improvement instructions for the
next generation. Loops until score meets threshold or max iterations.

Design doc:  ../../design.md
Overview:    ../../overview.md
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Protocol


# ── LLM interface ─────────────────────────────────────────────────────────────

class LLM(Protocol):
    def generate(self, messages: list[dict]) -> str: ...


# ── Core types ────────────────────────────────────────────────────────────────

@dataclass
class Iteration:
    number: int
    output: str
    score: float         # 0.0 – 1.0
    feedback: str
    passed: bool


@dataclass
class EvalResult:
    final_output: str
    passed: bool
    iterations: list[Iteration] = field(default_factory=list)
    final_score: float = 0.0


# ── Implementation ────────────────────────────────────────────────────────────

class EvaluatorOptimizer:
    """
    Iteratively improves LLM output via an evaluate-then-optimize loop.

    Usage:
        eo = EvaluatorOptimizer(generator_llm, evaluator_llm, criteria="...")
        result = eo.run("Write a haiku about databases")
    """

    EVALUATE_PROMPT = """\
Evaluate the following output against these criteria:
{criteria}

Output to evaluate:
{output}

Respond in this exact format:
SCORE: <number from 0.0 to 1.0>
FEEDBACK: <specific, actionable feedback for improvement>
PASS: <yes or no>"""

    IMPROVE_PROMPT = """\
Improve the following output based on the feedback provided.

Original task: {task}
Current output: {output}
Feedback: {feedback}

Produce an improved version that addresses all feedback points."""

    def __init__(
        self,
        generator: LLM,
        evaluator: LLM,
        criteria: str,
        threshold: float = 0.8,
        max_iterations: int = 3,
    ):
        self.generator = generator
        self.evaluator = evaluator
        self.criteria = criteria
        self.threshold = threshold
        self.max_iterations = max_iterations

    def _generate(self, task: str, previous: str | None, feedback: str | None) -> str:
        if previous is None:
            messages = [{"role": "user", "content": task}]
        else:
            messages = [{"role": "user", "content": self.IMPROVE_PROMPT.format(
                task=task, output=previous, feedback=feedback
            )}]
        return self.generator.generate(messages)

    def _evaluate(self, output: str) -> tuple[float, str, bool]:
        messages = [{"role": "user", "content": self.EVALUATE_PROMPT.format(
            criteria=self.criteria, output=output
        )}]
        raw = self.evaluator.generate(messages)

        # Parse structured response
        score, feedback, passed = 0.5, raw, False
        for line in raw.splitlines():
            if line.startswith("SCORE:"):
                try:
                    score = float(line.split(":", 1)[1].strip())
                except ValueError:
                    pass
            elif line.startswith("FEEDBACK:"):
                feedback = line.split(":", 1)[1].strip()
            elif line.startswith("PASS:"):
                passed = line.split(":", 1)[1].strip().lower() == "yes"

        return score, feedback, passed

    def run(self, task: str) -> EvalResult:
        output: str | None = None
        feedback: str | None = None
        iterations: list[Iteration] = []

        for i in range(self.max_iterations):
            output = self._generate(task, output, feedback)
            score, feedback, passed = self._evaluate(output)

            iterations.append(Iteration(
                number=i + 1,
                output=output,
                score=score,
                feedback=feedback,
                passed=passed,
            ))

            if passed or score >= self.threshold:
                return EvalResult(
                    final_output=output,
                    passed=True,
                    iterations=iterations,
                    final_score=score,
                )

        # Exhausted iterations — return best attempt
        return EvalResult(
            final_output=output or "",
            passed=False,
            iterations=iterations,
            final_score=iterations[-1].score if iterations else 0.0,
        )


# ── Example ───────────────────────────────────────────────────────────────────

if __name__ == "__main__":

    class MockLLM:
        def __init__(self, name: str):
            self.name = name
            self._call_count = 0

        def generate(self, messages: list[dict]) -> str:
            self._call_count += 1
            content = messages[-1]["content"]
            if self.name == "evaluator":
                score = min(0.6 + self._call_count * 0.2, 1.0)
                passed = "yes" if score >= 0.8 else "no"
                return f"SCORE: {score:.1f}\nFEEDBACK: Add more detail\nPASS: {passed}"
            return f"[v{self._call_count}] Generated output for: {content[:40]}"

    eo = EvaluatorOptimizer(
        generator=MockLLM("generator"),
        evaluator=MockLLM("evaluator"),
        criteria="Must be clear, accurate, and under 100 words.",
        threshold=0.8,
        max_iterations=3,
    )

    result = eo.run("Explain what a transformer neural network is")
    print(f"Passed:  {result.passed}")
    print(f"Iterations: {len(result.iterations)}")
    print(f"Final score: {result.final_score:.2f}")
    print(f"Output:\n{result.final_output}")

File: patterns/evaluator-optimizer/code/typescript/vercel-ai-sdk/evaluator-optimizer.ts
/**
 * Evaluator-Optimizer — Vercel AI SDK variant.
 *
 * Pattern: Evaluator-Optimizer (generator produces output; evaluator scores
 *   it against explicit criteria; if the score is below threshold the
 *   generator improves against the feedback; loops to a cap).
 * Framework: Vercel AI SDK (ai ^4.0.0) + @ai-sdk/anthropic (^1.0.0) + zod.
 * Idioms: generateText() for the generator's turns; generateObject() with
 *   a zod EvalSchema for the evaluator's turn so the score / feedback /
 *   pass tuple comes back typed instead of free-text the caller must parse.
 * Design doc: ../../../design.md (the framework-agnostic
 *   ../../python/evaluator_optimizer.py runs the same loop until the
 *   evaluator score crosses the threshold or max_iterations).
 *
 * Install:  pnpm add ai @ai-sdk/anthropic zod
 *           (or via tsx-on-demand: npx tsx evaluator-optimizer.ts)
 * Run:      ANTHROPIC_API_KEY=... npx tsx evaluator-optimizer.ts
 *
 * Note: ESM only.
 * tsconfig snippet: { "module": "ESNext", "moduleResolution": "Bundler", "target": "ES2022" }
 */

import { anthropic } from "@ai-sdk/anthropic";
import { generateObject, generateText, type LanguageModelV1 } from "ai";
import { z } from "zod";

// ── Types ───────────────────────────────────────────────────────────────────

interface IterationRecord {
  number: number;
  output: string;
  score: number;
  feedback: string;
  passed: boolean;
}

interface EvalResult {
  finalOutput: string;
  passed: boolean;
  iterations: IterationRecord[];
  finalScore: number;
}

const EvalSchema = z.object({
  score: z.number().min(0).max(1),
  feedback: z.string(),
  passed: z.boolean(),
});

// ── Driver ──────────────────────────────────────────────────────────────────

class EvaluatorOptimizer {
  constructor(
    private readonly criteria: string,
    private readonly options: {
      threshold?: number;
      maxIterations?: number;
      generatorModel?: LanguageModelV1;
      evaluatorModel?: LanguageModelV1;
    } = {},
  ) {}

  private get threshold(): number {
    return this.options.threshold ?? 0.8;
  }

  private get maxIterations(): number {
    return this.options.maxIterations ?? 3;
  }

  private get generatorModel(): LanguageModelV1 {
    return this.options.generatorModel ?? anthropic("claude-sonnet-4-6");
  }

  private get evaluatorModel(): LanguageModelV1 {
    // Evaluator should outclass the generator where possible; falls back to
    // sonnet so a single API key configures both.
    return this.options.evaluatorModel ?? anthropic("claude-opus-4-7");
  }

  async run(task: string): Promise<EvalResult> {
    const iterations: IterationRecord[] = [];
    let previous: string | null = null;
    let lastFeedback: string | null = null;

    for (let i = 0; i < this.maxIterations; i++) {
      const output = await this.generate(task, previous, lastFeedback);
      const evaluation = await this.evaluate(output);
      const record: IterationRecord = {
        number: i + 1,
        output,
        score: evaluation.score,
        feedback: evaluation.feedback,
        passed: evaluation.passed,
      };
      iterations.push(record);

      if (record.passed || record.score >= this.threshold) {
        return {
          finalOutput: output,
          passed: true,
          iterations,
          finalScore: record.score,
        };
      }

      previous = output;
      lastFeedback = evaluation.feedback;
    }

    const last = iterations[iterations.length - 1];
    return {
      finalOutput: last?.output ?? "",
      passed: false,
      iterations,
      finalScore: last?.score ?? 0,
    };
  }

  private async generate(
    task: string,
    previous: string | null,
    feedback: string | null,
  ): Promise<string> {
    const prompt =
      previous === null
        ? task
        : `Original task: ${task}\nCurrent output: ${previous}\nFeedback: ${feedback}\n\nProduce an improved version that addresses every feedback point.`;
    const result = await generateText({
      model: this.generatorModel,
      prompt,
    });
    return result.text;
  }

  private async evaluate(output: string): Promise<z.infer<typeof EvalSchema>> {
    const result = await generateObject({
      model: this.evaluatorModel,
      schema: EvalSchema,
      system:
        "You are an evaluator. Score the output against the criteria, return concrete actionable feedback, and decide whether it passes.",
      prompt: `Criteria:\n${this.criteria}\n\nOutput:\n${output}`,
    });
    return result.object;
  }
}

// ── Smoke runner ────────────────────────────────────────────────────────────

async function main(): Promise<void> {
  if (!process.env.ANTHROPIC_API_KEY) {
    console.log(
      "Skipping smoke run — set ANTHROPIC_API_KEY to exercise the evaluate-optimise loop.",
    );
    return;
  }

  const eo = new EvaluatorOptimizer(
    "Must be clear, accurate, and under 100 words.",
    { threshold: 0.8, maxIterations: 3 },
  );

  const result = await eo.run("Explain what a transformer neural network is.");

  console.log(`Passed:      ${result.passed}`);
  console.log(`Iterations:  ${result.iterations.length}`);
  console.log(`Final score: ${result.finalScore.toFixed(2)}`);
  console.log(`Output:\n${result.finalOutput}`);
}

if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch((err) => {
    console.error(err);
    process.exit(1);
  });
}

export { EvaluatorOptimizer };
export type { IterationRecord, EvalResult };
